lib/src/embedded/dispatcher.ts

```diff
@@ -0,0 +1,332 @@
+// Copyright 2020 Google Inc. Use of this source code is governed by an
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import {Observable, Subject, Subscription} from 'rxjs';
+import {filter, map} from 'rxjs/operators';
+
+import {InboundMessage, OutboundMessage} from '../vendor/embedded_sass_pb';
+import {
+  InboundRequest,
+  InboundRequestType,
+  InboundResponse,
+  InboundResponseType,
+  InboundTypedMessage,
+  OutboundRequest,
+  OutboundRequestType,
+  OutboundResponse,
+  OutboundResponseType,
+  OutboundTypedMessage,
+} from './message-transformer';
+
+type PromiseOr<T> = T | Promise<T>;
+
+/**
+ * Dispatches requests, responses, and events.
+ *
+ * Consumers can send an inbound request. This returns a promise that resolves
+ * with the corresponding outbound response, or errors if any Protocol Errors
+ * were encountered.
+ *
+ * Consumers can also register callbacks for processing different types of
+ * outbound requests. When an outbound request arrives, this runs the
+ * appropriate callback to process it, then sends the result inbound. Errors
+ * are not thrown to outbound request listeners; outbound request streams die
+ * quietly.
+ *
+ * Requests and responses do not need to have their ID set, since the dispatcher
+ * handles setting the ID on all messages.
+ *
+ * Detects ProtocolErrors caused by mismatched requests and responses.
+ */
+export class Dispatcher {
+  // Tracks the IDs of all inbound requests. An outbound response with matching
+  // ID and type will remove the ID.
+  private readonly pendingInboundRequests = new RequestTracker();
+
+  // Tracks the IDs of all outbound requests. An inbound response with matching
+  // ID and type will remove the ID.
+  private readonly pendingOutboundRequests = new RequestTracker();
+
+  // Collects all errors that we might encounter.
+  private readonly error$ = new Subject<Error>();
```

Comment:
Thanks—that's way more robust.

Reply:
Would it make sense to also expose errors to all request listeners, instead of only request senders? In that case every listener would have to listen for and handle errors. (As opposed to current behavior of closing listener streams silently, and having them opt into listening for errors on the `error$` stream?)

Reply:
The problem is that there's not a very good way of exposing errors to the listeners given that they're just passing in a callback. `Observable`s have all this nice machinery around error handling like `catchError()` and so on, but a callback is just a callback.

Reply:
We could return the entire `Observable` to the listener when they register their callback, instead of just the subscription

Reply:
But that doesn't give them a built-in way to send a reply back. The nice thing about the callback is that it's a type-safe encapsulation of the contract to give one response for each request the caller receives.

Reply:
Right—what I meant to say is, we keep the callback registration, but also return the observable so the listener can choose to perform error handling. For example,

```ts
// dispatcher
onImportRequest(): Observable

// listener
const importRequests = dispatcher.onImportRequest(() => {});
importRequests.subscribe( /* handle error */ );
```

vs.

```ts
// dispatcher
onImportRequest(): Subscription

// listener
const importRequests = dispatcher.onImportRequest(() => {});
dispatcher.error$.subscribe( /* handle error */ );
```

Reply:
If you do that, then *every* caller is obligated to handle errors for *every* stream, otherwise they risk leaving unhandled errors.

Also, you'd *still* want an `error$` stream so that you could access errors if you spin up a dispatcher without subscribing to any event streams.